src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -559,5 +621,529 @@ static void Main()
+  IL_0082:  call       ""string string.Concat(object, object, object)""
+  IL_0087:  call       ""void System.Console.WriteLine(string)""
+  IL_008c:  ret
+}
+");
+        }
+
+        [Fact(Skip = "PROTOTYPE(tuples)")]
+        public void DeconstructLongTuple2()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        long x;
+        int y;
+
+        (x, x, x, x, x, x, x, x, x, y) = (1, 1, 1, 1, 1, 1, 1, 1, 4, (byte)2);
+        System.Console.WriteLine(x + "" "" + y);
+    }
+}
+";
+
+            // issue with return type
+            var comp = CompileAndVerify(source, expectedOutput: "4 2", additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular.WithTuplesFeature());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void DeconstructTypelessTuple()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        string x = ""goodbye"";
+        string y;
+
+        (x, y) = (null, ""hello"");
+        System.Console.WriteLine($""{x}{y}"");
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "hello", additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular.WithTuplesFeature());
+            comp.VerifyDiagnostics();
+            comp.VerifyIL("C.Main", @"
+{
+  // Code size       48 (0x30)
+  .maxstack  3
+  .locals init (string V_0, //x
+                string V_1) //y
+  IL_0000:  ldstr      ""goodbye""
+  IL_0005:  stloc.0
+  IL_0006:  ldnull
+  IL_0007:  ldstr      ""hello""
+  IL_000c:  newobj     ""System.ValueTuple<string, string>..ctor(string, string)""
+  IL_0011:  dup
+  IL_0012:  ldfld      ""string System.ValueTuple<string, string>.Item1""
+  IL_0017:  stloc.0
+  IL_0018:  ldfld      ""string System.ValueTuple<string, string>.Item2""
+  IL_001d:  stloc.1
+  IL_001e:  ldstr      ""{0}{1}""
+  IL_0023:  ldloc.0
+  IL_0024:  ldloc.1
+  IL_0025:  call       ""string string.Format(string, object, object)""
+  IL_002a:  call       ""void System.Console.WriteLine(string)""
+  IL_002f:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TupleWithNoConversion()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        byte x;
+        string y;
+
+        (x, y) = (1, 2);
+    }
+}
+";
+            // PROTOTYPE(tuples) This error message is misleading
+            var comp = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular.WithTuplesFeature());
+            comp.VerifyDiagnostics(
+                // (9,18): error CS0029: Cannot implicitly convert type '(int, int)' to '(byte, string)'
+                //         (x, y) = (1, 2);
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "(1, 2)").WithArguments("(int, int)", "(byte, string)").WithLocation(9, 18)
+                );
+        }
```

Comment:
Add tests where RHS is used on the left:

```C#
(object i, object ii) x = (1,2);
object y;

(x.ii, y) = x;
```

should result in    x: {1, 1}  , y: 2